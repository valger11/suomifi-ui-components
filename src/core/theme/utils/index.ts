export { margin, padding, spacingModifiers } from './spacing';
