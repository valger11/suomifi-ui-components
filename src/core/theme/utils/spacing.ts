import {
  SuomifiTheme,
  SpacingProp,
  SpacingDesignTokens,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

type SpacingDesignTokensKeys = keyof SpacingDesignTokens;

export const spacingTokensKeys = Object.keys(
  defaultSuomifiTheme.spacing,
) as SpacingDesignTokensKeys[];

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

/**
 * Spacing type
 * @typedef {String} spacingType
 * @param {String} type 'padding' or 'margin'
 */
/**
 * Spacing tokens
 * @typedef {String} spacingTokens
 * @param {String} t token for top, or top and bottom, or all
 * @param {String} r token for right, or right and left
 * @param {String} b token for bottom
 * @param {String} l token for left
 */
/**
 * Set margin or padding pased on tokens
 * @param {Object} theme suomifiTheme
 * @return {(spacingType) => (spacingTokens) => String}
 */
const space = (theme: SuomifiTheme) => (type: 'padding' | 'margin') => (
  t?: SpacingProp,
  r?: SpacingProp,
  b?: SpacingProp,
  l?: SpacingProp,
) =>
  !!t
    ? `${!!t ? `${type}-top: ${spaceVal(theme)(t)};` : ''}
    ${!!r ? `${type}-right: ${spaceVal(theme)(r)};` : ''}
    ${!!b ? `${type}-bottom: ${spaceVal(theme)(b)};` : ''}
    ${!!l ? `${type}-left:${spaceVal(theme)(l)};` : ''}`
    : '';

/**
 * Set margin based on theme
 * @param {Object} theme
 */
export const margin = (theme: SuomifiTheme) => space(theme)('margin');
/**
 * Set padding based on theme
 * @param {Object} theme
 */
export const padding = (theme: SuomifiTheme) => space(theme)('padding');

/**
 * Create spacing styles for CSS-selector (-xxs, -xs, -s...)
 * TODO: this should be in suomifiTheme? or integrated somehow with that
 * @param {Object} tokens Design tokens
 * @return {(spacingType) => (selector: String) => String}
 */
export const spacingModifiers = (theme: SuomifiTheme) => (
  spacing: 'padding' | 'margin' | 'margin-bottom',
) => (selector: string) =>
  spacingTokensKeys.reduce(
    (ret, k) =>
      `${ret}${selector}-${k}{
        ${spacing}: ${spaceVal(theme)(k)}
  } `,
    '',
  );
