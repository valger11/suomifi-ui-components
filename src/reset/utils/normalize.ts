export const normalize = {
  html:
    'line-height: 1.15; -ms-text-size-adjust: 100%; -webkit-text-size-adjust: 100%;',
  h1: 'font-size: 2em; margin: 0.67em 0;',
  'dl dl': 'margin: 0;',
  'dl ol': 'margin: 0;',
  'dl ul': 'margin: 0;',
  'ol dl': 'margin: 0;',
  'ul dl': 'margin: 0;',
  'ol ol': 'margin: 0;',
  'ol ul': 'margin: 0;',
  'ul ol': 'margin: 0;',
  'ul ul': 'margin: 0;',
  hr: 'box-sizing: content-box; height: 0; overflow: visible;',
  main: 'display: block;',
  pre: 'font-family: monospace, monospace; font-size: 1em;',
  a: 'background-color: transparent;',
  'abbr[title]':
    'text-decoration: underline; text-decoration: underline dotted;',
  b: 'font-weight: bolder;',
  strong: 'font-weight: bolder;',
  code: 'font-family: monospace, monospace; font-size: 1em;',
  kbd: 'font-family: monospace, monospace; font-size: 1em;',
  samp: 'font-family: monospace, monospace; font-size: 1em;',
  small: 'font-size: 80%;',
  audio: 'display: inline-block;',
  video: 'display: inline-block;',
  'audio:not([controls])': 'display: none; height: 0;',
  img: 'border-style: none;',
  'svg:not(:root)': 'overflow: hidden;',
  button:
    'margin: 0;overflow: visible; text-transform: none;-webkit-appearance: button;',
  input: 'margin: 0;overflow: visible;',
  select: 'margin: 0;text-transform: none;',
  '[type="button"]': '-webkit-appearance: button;',
  '[type="reset"]': '-webkit-appearance: button;',
  '[type="submit"]': '-webkit-appearance: button;',
  fieldset: 'padding: 0.35em 0.75em 0.625em;',
  legend:
    'box-sizing: border-box; color: inherit; display: table; max-width: 100%; white-space: normal;',
  progress: 'display: inline-block; vertical-align: baseline;',
  textarea: 'margin: 0; overflow: auto;',
  '[type="checkbox"]': 'box-sizing: border-box; padding: 0;',
  '[type="radio"]': 'box-sizing: border-box; padding: 0;',
  '[type="search"]': '-webkit-appearance: textfield; outline-offset: -2px;',
  '::-webkit-inner-spin-button': 'height: auto;',
  '::-webkit-outer-spin-button': 'height: auto;',
  '::-webkit-input-placeholder': 'color: inherit; opacity: 0.54;',
  '::-webkit-search-decoration': '-webkit-appearance: none;',
  '::-webkit-file-upload-button': '-webkit-appearance: button; font: inherit;',
  '::-moz-focus-inner': 'border-style: none; padding: 0;',
  ':-moz-focusring': 'outline: 1px dotted ButtonText;',
  ':-moz-ui-invalid': 'box-shadow: none;',
  details: 'display: block;',
  dialog:
    // eslint-disable-next-line max-len
    'background-color: white; border: solid; color: black; display: block; height: -moz-fit-content; height: -webkit-fit-content; height: fit-content; left: 0; margin: auto; padding: 1em; position: absolute; right: 0; width: -moz-fit-content; width: -webkit-fit-content; width: fit-content;',
  'dialog:not([open])': 'display: none;',
  summary: 'display: list-item;',
  canvas: 'display: inline-block;',
  template: 'display: none;',
  '[hidden]': 'display: none;',
} as { [key: string]: string };
