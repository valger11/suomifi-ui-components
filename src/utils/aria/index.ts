export {
  ariaLabelOrHidden,
  ariaFocusableNoLabel,
  getConditionalAriaProp,
} from './aria';
