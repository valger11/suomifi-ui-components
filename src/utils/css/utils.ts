import { css, FlattenSimpleInterpolation } from 'styled-components';
import { cssValueToString } from './cssvalue';

export const clearfix = css`
  &:after {
    display: block;
    content: '';
    clear: both;
  }
`;

const camelToSnake = (string: string) =>
  string.replace(/[\w]([A-Z])/g, (m) => `${m[0]}-${m[1]}`).toLowerCase();
/**
 * Convert CSSObject to CSS FlattenSimpleInterpolation
 * @param value CSSObject
 *
 * @example
 *  cssObjectToCss({
 *    fontFamily: "'Arial', sans-serif";
 *    fontSize: {value: 16, unit: 'px'};
 *  })
 */
export const cssObjectToCss = <T>(object: T) => css`
  ${Object.entries(object)
    .map(([key, value]) => `${camelToSnake(key)}: ${cssValueToString(value)};`)
    .join('')}
`;

/**
 * Convert CSSObject tokens to CSS FlattenSimpleInterpolation
 * @param tokens Tokens of CSSObjects
 *
 * @example
 *  cssObjectsToCss({
 *    bodyText: {
 *      fontFamily: "'Arial', sans-serif";
 *      fontSize: {value: 16, unit: 'px'};
 *    },
 *    bodyTextSmallScreen: {
 *      fontFamily: "'Arial', sans-serif";
 *      fontSize: {value: 18, unit: 'px'};
 *     }
 *  })
 */
export const cssObjectsToCss = <T, K extends keyof T>(tokens: T) =>
  Object.entries(tokens).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: cssObjectToCss(value),
    }),
    {} as { [key in K]: FlattenSimpleInterpolation },
  );
