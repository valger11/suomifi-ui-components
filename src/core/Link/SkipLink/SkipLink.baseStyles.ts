import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { baseStyles } from '../BaseLink/BaseLink.baseStyles';

export const SkipLinkStyles = (theme: SuomifiTheme) => css`
  ${baseStyles(theme)}
  &.fi-link--skip {
    position: absolute;
    z-index: 10000;
    left: -1000px;
    margin: ${theme.spacing.insetXl};
    padding: ${theme.spacing.insetM};
    background: ${theme.colors.highlightLight3};
    border: 1px solid ${theme.colors.depthLight1};
    color: ${theme.colors.blackBase};
    text-decoration: none;
  }

  &.fi-link--skip:focus {
    position: absolute;
    left: auto;
  }
`;
