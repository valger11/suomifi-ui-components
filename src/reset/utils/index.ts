import { css } from 'styled-components';
import { normalize } from './normalize';

const common = css`
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;

  ::before,
  ::after {
    box-sizing: border-box;
  }
`;

const resets = {
  normalize,
  common,
};

const selectorDeclaration = (selector: string, styles: string) =>
  `&${selector} {
    ${styles}
  }`;

const selectorsAndCss = (selector: string) => {
  const styles = !!normalize && normalize[selector];
  return !!styles ? selectorDeclaration(selector, styles) : '';
};

/**
 * Return selectors with resets in Emotion-css
 * @param selectors string or array of strings
 */
const resetWithSelectors = (selectors: string | string[]): string => {
  if (Array.isArray(selectors) && selectors.length > 0) {
    return `
      ${selectors.map((s) => selectorsAndCss(s)).join('')}
    `;
  }
  if (typeof selectors === 'string') {
    return selectorsAndCss(selectors);
  }
  return '';
};

export { resets, resetWithSelectors };

// TODO: Add tests
