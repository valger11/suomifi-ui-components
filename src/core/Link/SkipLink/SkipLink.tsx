import React, { Component } from 'react';
import classnames from 'classnames';
import { default as styled } from 'styled-components';
import { Link } from '../Link/Link';
import { SkipLinkStyles } from './SkipLink.baseStyles';
import { BaseLinkProps } from '../BaseLink/BaseLink';
import { SuomifiTheme, SuomifiThemeConsumer } from '../../theme';

const skipClassName = 'fi-link--skip';

export interface SkipLinkProps extends BaseLinkProps {}

const StyledSkipLink = styled(
  (props: SkipLinkProps & { theme: SuomifiTheme }) => {
    const { theme, ...passProps } = props;
    return <Link {...passProps} />;
  },
)`
  ${({ theme }) => SkipLinkStyles(theme)}
`;

/**
 * <i class="semantics" />
 * Used for adding skip link for keyboard and screenreader users
 */
export class SkipLink extends Component<SkipLinkProps> {
  render() {
    const { className, ...passProps } = this.props;
    return (
      <SuomifiThemeConsumer>
        {({ suomifiTheme }) => (
          <StyledSkipLink
            theme={suomifiTheme}
            {...passProps}
            className={classnames(className, skipClassName)}
          />
        )}
      </SuomifiThemeConsumer>
    );
  }
}
