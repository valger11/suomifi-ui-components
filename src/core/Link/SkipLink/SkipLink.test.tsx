import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test';

import { SkipLink } from './SkipLink';

const TestSkipLink = (
  <SkipLink href="/" data-testid="test-link">
    Test content
  </SkipLink>
);

test('should match snapshot', () => {
  const LinkRendered = render(TestSkipLink);
  const { getByTestId, container } = LinkRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('test-link').textContent).toBe('Test content');
});

test('should not have basic accessibility issues', axeTest(TestSkipLink));
