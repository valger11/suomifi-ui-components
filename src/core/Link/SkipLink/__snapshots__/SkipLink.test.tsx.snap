// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c0:hover,
.c0:active,
.c0:focus,
.c0:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c1 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:hover,
.c1:active,
.c1:focus,
.c1:focus-within {
  color: hsl(212,63%,45%);
}

.c1:focus,
.c1:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:focus {
  outline: 0;
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
}

.c1:hover,
.c1:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:visited {
  color: hsl(284,36%,45%);
}

.c2 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover,
.c2:active,
.c2:focus,
.c2:focus-within {
  color: hsl(212,63%,45%);
}

.c2:focus,
.c2:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:focus {
  outline: 0;
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
}

.c2:hover,
.c2:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c2:visited {
  color: hsl(284,36%,45%);
}

.c2.fi-link--skip {
  position: absolute;
  z-index: 10000;
  left: -1000px;
  margin: 16px;
  padding: 8px;
  background: hsl(212,63%,95%);
  border: 1px solid hsl(202,7%,80%);
  color: hsl(0,0%,16%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2.fi-link--skip:focus {
  position: absolute;
  left: auto;
}

<a
  class="c0 fi-link c1 c2 fi-link--skip"
  data-testid="test-link"
  href="/"
>
  Test content
</a>
`;
